### ISSUE TITLE
destroy functions should take references to the object being destroyed and should re-initialize the reference

### ISSUE THREAD

**Issue by jherico** [nvpro-pipeline/vkcpp#114]
consider 
```
    void Device::destroyImageView( ImageView imageView, Optional<const AllocationCallbacks> allocator = nullptr ) const
    {
      vkDestroyImageView( m_device, static_cast<VkImageView>( imageView ), reinterpret_cast<const VkAllocationCallbacks*>( static_cast<const AllocationCallbacks*>( allocator)) );
    }
```

It would be valuable if this was instead 
```
    void Device::destroyImageView( ImageView& imageView, Optional<const AllocationCallbacks> allocator = nullptr ) const
    {
      vkDestroyImageView( m_device, static_cast<VkImageView>( imageView ), reinterpret_cast<const VkAllocationCallbacks*>( static_cast<const AllocationCallbacks*>( allocator)) );
      imageView = ImageView();
    }
```

re-initializing the object to a null handle makes it very obvious to the program that a resource has been freed.

**Comment by asuessenbach** [comment]
I agree, resetting the destroyed class would be a valuable approach.

The hard part here is, to determine that something is supposed to be destroyed from the information available in vk.xml. Is it the substring "Destroy" in the function name? Or "Free"? And is the string following "Destroy" the type of the thing to be destroyed? Is there an argument of that same type available?

You see, it's not that obvious how to determine that.
And following the "destroy"-rule might lead to different handling of different stuff, thus leading to unexpected inconsistencies.
And maybe it's not even what you want, as there's some possibly unexpected additional operation done, which takes some time?

Any opinions on that?

**Comment by jherico** [comment]
@asuessenbach I'm not convinced that there's not a heuristic for determining all the methods that should reset their inputs, but even if there's not, I think the solution to that is to find a heuristic that covers the vast majority of the cases and either 

- hand code support for the rest in the generator
- allow the generator to take as input an XSL stylesheet that can transform and annotate various functions with additional processing flags.  This prevents you from having to put special case logic in the code, but allows you to continue to use the upstream vk.xml.
